keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,577 @@
+proxy the traffic to, otherwise it return an “Unknown network” error. 
+
+The KAS starts with a proxy configuration like the below example. 
+The example specifies 4 networks. "direct" specifies the KAS talking directly 
+on the local network (no proxy). "master" specifies the KAS talks to a proxy
+listening at 1.2.3.4:5678. "cluster" specifies the KAS talk to a proxy 
+listening at 1.2.3.5:5679. While these are represented as resources 
+they are not intended to be loaded dynamically. 
+The KAS loads them as configuration at start time.
+
+```yaml
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: direct
+  connection:
+    type: direct
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Master
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5678
+    caBundle: file1.ca
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Cluster
+  connection:
+    type: grpc
+    url: grpc://1.2.3.5:5679
+    caBundle: file2.ca
+
+```
+
+NetworkContext could be extended to contain more contextual information.
+This would allow smarter routing based on the k8s object KAS is processing 
+or which user/tenant tries to initiate the request, etc.
+
+### Proxy gRPC definition
+
+```grpc
+syntax = "proto3";
+
+service ProxyService {
+    // Proxy connects to a remote address by stream id, and establish
+    // a bi-directional stream of packet exchange.
+    rpc Proxy(stream Packet) returns (stream Packet) {}
+}
+
+enum PacketType {
+  DIAL_REQ = 0;
+  DIAL_RSP = 1;
+  CLOSE_REQ = 2;
+  CLOSE_RSP = 3;
+  DATA = 4;
+}
+
+message Packet {
+  PacketType type = 1;
+
+  oneof payload {
+    DialRequest dialRequest = 2;
+    DialResponse dialResponse = 3;
+    CloseRequest closeRequest = 4;
+    CloseResponse closeResponse = 5;
+    Data data = 6;
+  }
+}
+
+// Error is sent when error happens from remote side of connection when it tries
+// to read or write from it
+enum Error {
+  EOF = 0;
+  EIO = 1;
+  ECONNRESET = 2;
+  ETIMEOUT = 3;
+  EADDRNOTAVAIL = 4;
+  EMFILE = 5;
+  // … 
+}
+
+// DialRequest represents a request to dial to an address on the other 
+// side of the tunnel. The format is inspired by golang's net interface
+// https://golang.org/pkg/net/#Dial 
+message DialRequest {
+    // network representing a named network. "Tcp" is the only supported
+    // value
+    string network = 1;
+
+    // For TCP network, the address has the form "ip:port", where host must
+    // be IP address.
+    string address = 2;
+
+    // random is the randomly generated bytes that represents a given dial request.
+    // the number is kept unchanged across the proxies, and copied to DialResponse
+    // from the other side of the tunnel. The number cannot be reused across 
+    // different DialRequests.
+    bytes random = 3;
```

Comment:
How is this used? Why is the grpc stream id not sufficient?

Reply:
This is to multiplex more than one connections through the same grpc stream. Once a grpc stream is established, packets can flow back and forth between both ends of the stream. The also could be multiple dial requests sent out at the same time, random is used to match between response and request.

Reply:
I guessed that, but why is it useful to multiplex connections over a single grpc stream? Grpc streams are already multiplexed over a single tcp connection.

Reply:
I see what you mean. Good question. Actually there is another option to avoid this, but it would be 2 method calls. Mostly because the grpc stream doesn't allow you to specify a destination. so you have to Dail then Connect. It's similar (not exactly) to socket, where you need to call socket to get a fd then connect to get another one for streaming. 
```proto
service ProxyService {
   // DialRequest and DialResponse is the same, but without random.
   rpc Dial(req DialRequest) DialResponse {}

   rpc Connect(stream Packet) (stream Packet) {}
}
```

This potentially moves the stream control out, and we probably need another method `Close` as well. The problem of this approach is that it's hard to define the order of multiple data and control flows in different serving threads. e.g. say when a close request and data comes in at the same time, it's hard to tell who's coming first. The right behavior is pretty much hard to define. While in the same stream, it's all sequential, then it's just a matter of whoever arrives first.

Reply:
> The problem of this approach is that it's hard to define the order of multiple data and control flows in different serving threads. e.g. say when a close request and data comes in at the same time, it's hard to tell who's coming first. The right behavior is pretty much hard to define. While in the same stream, it's all sequential, then it's just a matter of whoever arrives first.

Assume packets arrive reliably and in order, send dial parameters in a message extension or header and you have yourself a drop in replacement for net.Dial with all the hard stuff handled by the lower layers :)